pkg/eestream/encode.go

```diff
@@ -38,71 +38,103 @@ type ErasureScheme interface {
 }
 
 type encodedReader struct {
-	r               io.Reader
-	es              ErasureScheme
-	cv              *sync.Cond
-	inbuf           []byte
-	outbufs         [][]byte
-	piecesRemaining int
-	err             error
+	r       io.Reader
+	es      ErasureScheme
+	inbuf   []byte
+	outbufs [][]byte
+	chans   map[int]chan []byte
+	err     error
 }
 
 // EncodeReader will take a Reader and an ErasureScheme and return a slice of
 // Readers
-func EncodeReader(r io.Reader, es ErasureScheme) []io.Reader {
+func EncodeReader(r io.Reader, es ErasureScheme, maxBufferMemory int) []io.Reader {
 	er := &encodedReader{
 		r:       r,
 		es:      es,
-		cv:      sync.NewCond(&sync.Mutex{}),
 		inbuf:   make([]byte, es.DecodedBlockSize()),
 		outbufs: make([][]byte, es.TotalCount()),
+		chans:   make(map[int]chan []byte, es.TotalCount()),
 	}
 	readers := make([]io.Reader, 0, es.TotalCount())
 	for i := 0; i < es.TotalCount(); i++ {
-		er.outbufs[i] = make([]byte, 0, es.EncodedBlockSize())
 		readers = append(readers, &encodedPiece{
 			er: er,
 			i:  i,
 		})
 	}
+	if maxBufferMemory < 0 {
+		er.err = Error.New("negative max buffer memory")
+		return readers
+	}
+	chanSize := maxBufferMemory / (es.TotalCount() * es.EncodedBlockSize())
+	if chanSize < 1 {
+		chanSize = 1
+	}
+	for i := 0; i < es.TotalCount(); i++ {
+		er.chans[i] = make(chan []byte, chanSize)
+	}
+	go er.fillBuffer()
```

Comment:
do we/should we care about the errors returned?

Reply:
Yes, we do, but not in the constructor. The constructor should return quickly, thus filling the buffer in a goroutine. If an error occur during filling the buffer, the goroutine will close all channels and exit. This will close all readers returned by the constructor. Perhaps, an improvement would be that the goroutine sends a message to the channels with the exact error that occurred before closing them.